dll/win32/shell32/shlexec.cpp

```diff
@@ -777,55 +777,33 @@ static UINT SHELL_FindExecutable(LPCWSTR lpPath, LPCWSTR lpFile, LPCWSTR lpVerb,
     }
 
     GetCurrentDirectoryW(ARRAY_SIZE(curdir), curdir);
-    if (!PathIsFileSpecW(lpFile))
+    if (lpPath && *lpPath)
     {
-        BOOL found = FALSE;
-        if (lpPath && *lpPath)
-        {
-            TRACE("lpPath %s\n", debugstr_w(lpPath));
-            PathCombineW(xlpFile, lpPath, lpFile);
-            if (PathFileExistsDefExtW(xlpFile, WHICH_DEFAULT | WHICH_OPTIONAL) || PathFileExistsW(xlpFile))
-            {
-                GetFullPathNameW(xlpFile, ARRAY_SIZE(xlpFile), xlpFile, NULL);
-                found = TRUE;
-            }
-        }
-        if (!found)
-        {
-            lstrcpyW(xlpFile, lpFile);
-            if (PathFileExistsDefExtW(xlpFile, WHICH_DEFAULT | WHICH_OPTIONAL) || PathFileExistsW(xlpFile))
-            {
-                GetFullPathNameW(xlpFile, ARRAY_SIZE(xlpFile), xlpFile, NULL);
-                found = TRUE;
-            }
-        }
-        if (found)
-        {
-            lpFile = xlpFile;
-            lstrcpyW(lpResult, xlpFile);
-        }
-        else
-            xlpFile[0] = '\0';
+        search_paths[0] = lpPath;
+        search_paths[1] = curdir;
     }
     else
+        search_paths[0] = curdir;
+    lstrcpyW(xlpFile, lpFile);
+    if (PathResolveW(xlpFile, search_paths, PRF_TRYPROGRAMEXTENSIONS | PRF_VERIFYEXISTS) ||
+        PathFindOnPathW(xlpFile, search_paths))
     {
-        if (lpPath && *lpPath)
-        {
-            search_paths[0] = lpPath;
-            search_paths[1] = curdir;
-        }
-        else
-            search_paths[0] = curdir;
-        lstrcpyW(xlpFile, lpFile);
-        if (PathResolveW(xlpFile, search_paths, PRF_TRYPROGRAMEXTENSIONS | PRF_VERIFYEXISTS))
-        {
-            TRACE("PathResolveW returned non-zero\n");
-            lpFile = xlpFile;
-            lstrcpyW(lpResult, xlpFile);
-            /* The file was found in lpPath or one of the directories in the system-wide search path */
-        }
+        TRACE("PathResolveW returned non-zero\n");
+        lpFile = xlpFile;
+        lstrcpyW(lpResult, xlpFile);
+        /* The file was found in lpPath or one of the directories in the system-wide search path */
+    }
+    else
+        xlpFile[0] = '\0';
+
+    /* Clear any trailing spaces */
```

Comment:
Can the xlpFile contain _leading_ spaces though? If so and if both leading and trailing spaces can be trimmed, I would suggest you use `PathRemoveBlanksW(xlpFile);`

Reply:
Indeed. The correct path shouldn't contain neither leading nor trailing whitespaces at all (only in between of the parts, e. g. `part 1\part 2`). But if to consider a possibility of a wrong (test)case, such a path actually can exist. So yes, the function you've suggested looks a better solution for this.
But I wonder can this function be useful when only one usage case is required (the path contains either leading or trailing whitespaces **only**)? :thinking:

Reply:
Applied, thanks.

Reply:
I haven't found an existing shell/win32 function that only does leading, or only trailing whitespace trimming.
If your string was part of an ATL CString object, you could use TrimLeft or TrimRight (but not that these don't only remove space `' '` characters, but also newlines and tabs). In the C++ stl there are similar functions (for std::string and similar objects). Not sure there's a function in the C standard library: you would have to do a similar loop as yours